ddtrace/llmobs/_integrations/base_stream_handler.py

```diff
@@ -0,0 +1,231 @@
+        except Exception as e:
+            self._exception = e
+            self._handler.handle_exception(e)
+            raise
+        finally:
+            self._handler.finalize_stream(self._exception)
+    
+    def __enter__(self):
+        if hasattr(self.__wrapped__, '__enter__'):
+            result = self.__wrapped__.__enter__()
+            # update iterator in case we are wrapping a stream manager
+            if result is not self.__wrapped__:
+                self._stream_iter = iter(result) if not hasattr(result, '__next__') else result
+                new_stream = _ClassTracedStream(result, self._handler, self._on_stream_created)
+                if self._on_stream_created:
+                    self._on_stream_created(new_stream)
+                return new_stream
+        return self
+    
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if hasattr(self.__wrapped__, '__exit__'):
+            return self.__wrapped__.__exit__(exc_type, exc_val, exc_tb)
+    
+    @property
+    def handler(self):
+        return self._handler
+
+
+class _ClassTracedAsyncStream(wrapt.ObjectProxy):
+    def __init__(self, wrapped_stream, handler: AsyncStreamHandler, on_stream_created=None):
+        super().__init__(wrapped_stream)
+        self._handler = handler
+        self._on_stream_created = on_stream_created
+        self._exception = None
+        if hasattr(self.__wrapped__, '__anext__') or not hasattr(self.__wrapped__, '__aiter__'):
+            self._async_stream_iter = self.__wrapped__
+        else:
+            self._async_stream_iter = aiter(self.__wrapped__)
+    
+    async def __aiter__(self):
+        try:
+            async for chunk in self._async_stream_iter:
+                await self._handler.process_chunk(chunk, self._async_stream_iter)
+                yield chunk
+        except Exception as e:
+            self._exception = e
+            self._handler.handle_exception(e)
+            raise
+        finally:
+            self._handler.finalize_stream(self._exception)
+    
+    async def __aenter__(self):
+        if hasattr(self.__wrapped__, '__aenter__'):
+            result = await self.__wrapped__.__aenter__()
+            # update iterator in case we are wrapping a stream manager
+            if result is not self.__wrapped__:
+                self._async_stream_iter = aiter(result) if not hasattr(result, '__anext__') else result
+                new_stream = _ClassTracedAsyncStream(result, self._handler, self._on_stream_created)
+                if self._on_stream_created:
+                    self._on_stream_created(new_stream)
+                return new_stream
+        return self
+    
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        if hasattr(self.__wrapped__, '__aexit__'):
+            return await self.__wrapped__.__aexit__(exc_type, exc_val, exc_tb)
+    
+    @property
+    def handler(self):
+        return self._handler
+    
+
+class _GeneratorTracedStream:
```

Comment:
Why do we need separate classes for generators? My earlier `_self_*` comment might actually resolve any issues you might've had with using wrapt.ObjectProxy on generators, let me know if that helps